### PR TITLE
Iterative block tree implementation

### PR DESCRIPTION
**Pull request by Protryon**
This PR removes the recursive implementation of block tree fetching from sqlite in favor of an iterative approach -- not implemented as a SQL query for now.

### CODE REVIEW

File: storage/src/storage/sync.rs [hunk 1]
```diff
@@ -156,21 +156,68 @@ pub trait SyncStorage {
 
     /// Gets a tree structure representing all the descendents of [`block_hash`]
     fn get_block_digest_tree(&mut self, block_hash: &Digest) -> Result<DigestTree> {
-        let children = self.get_block_children(block_hash)?;
-        if children.is_empty() {
+        let mut nodes: HashMap<Digest, Vec<Digest>> = HashMap::new();
+        let mut stack = vec![block_hash.clone()];
+        while let Some(hash) = stack.pop() {
+            let children = self.get_block_children(&hash)?;
+            stack.extend_from_slice(&children[..]);
```
Comment (ljedrz) [inline storage/src/storage/sync.rs:163]: nit: this could be put in the conditional block right below as well

File: storage/src/storage/sync.rs [hunk 2]
```diff
@@ -156,21 +156,68 @@ pub trait SyncStorage {
...
-            let subtree = self.get_block_digest_tree(&child)?;
-            let len = subtree.longest_length() + 1;
-            if len > longest_tree_len {
-                longest_tree_len = len;
+        let mut node_entries: HashSet<Digest> = nodes.keys().cloned().collect();
+
+        let mut trees: HashMap<Digest, DigestTree> = HashMap::new();
+        loop {
+            if nodes.is_empty() {
+                break;
```
Comment (ljedrz) [inline storage/src/storage/sync.rs:176]: nit: `while !nodes.is_empty() {`

File: storage/src/storage/sync.rs [hunk 3]
```diff
@@ -156,21 +156,68 @@ pub trait SyncStorage {
...
+                trees.insert(
+                    hash.clone(),
+                    DigestTree::Node(hash.clone(), new_children, longest_tree_len),
+                );
+                node_entries.remove(hash);
+                false
+            });
         }
-        Ok(DigestTree::Node(block_hash.clone(), out_children, longest_tree_len))
+        assert_eq!(nodes.len(), 0);
```
Comment (ljedrz) [inline storage/src/storage/sync.rs:203]:
```suggestion
        assert_eq!(nodes.len(), 0);
```
already ensured by the breaking condition

File: storage/src/storage/sync.rs [hunk 4]
```diff
@@ -156,21 +156,68 @@ pub trait SyncStorage {
...
+        assert_eq!(trees.len(), 1);
```
Comment (ljedrz) [inline storage/src/storage/sync.rs:204]:
```suggestion
        debug_assert_eq!(trees.len(), 1);
```
unless it's severe enough to warrant a runtime panic?

File: storage/src/storage/sync.rs [hunk 5]
```diff
@@ -156,21 +156,68 @@ pub trait SyncStorage {
...
+        // let mut longest_tree_len = 0usize;
+        // for child in children {
+        //     let subtree = self.get_block_digest_tree(&child)?;
+        //     let len = subtree.longest_length() + 1;
+        //     if len > longest_tree_len {
+        //         longest_tree_len = len;
+        //     }
+        //     out_children.push(subtree);
+        // }
+        // Ok(DigestTree::Node(block_hash.clone(), out_children, longest_tree_len))
```
Comment (ljedrz) [inline storage/src/storage/sync.rs:220]:
```suggestion
```

File: storage/src/storage/sync.rs [hunk 6]
```diff
@@ -156,21 +156,68 @@ pub trait SyncStorage {
...
+                );
+                node_entries.remove(hash);
+                false
+            });
         }
-        Ok(DigestTree::Node(block_hash.clone(), out_children, longest_tree_len))
+        assert_eq!(nodes.len(), 0);
+        assert_eq!(trees.len(), 1);
+
+        Ok(trees.remove(block_hash).expect("missing block hash tree"))
```
Comment (ljedrz) [inline storage/src/storage/sync.rs:206]: again a question on panicking here; perhaps we could throw an error instead?

**Review by ljedrz — APPROVED** [review]
Left some comments, LGTM otherwise; going to test it a bit more locally, haven't run into issues yet

File: storage/src/storage/sync.rs (hunk 4 above)
Comment (Protryon) [reply storage/src/storage/sync.rs:204]: it is -- it implies a complete failure of the algorithm to manage a tree, which is an unrecoverable failure.

File: storage/src/storage/sync.rs (hunk 6 above)
Comment (Protryon) [reply storage/src/storage/sync.rs:206]: again, critical algorithmic failure -- not something that should be recovered from.

**Pull request merged**